gax/src/test/java/com/google/api/gax/retrying/ScheduledRetryingExecutorTest.java

```diff
@@ -47,74 +56,251 @@
+              .build();
+
+      RetryingExecutor<String> executor =
+          getRetryingExecutor(retrySettings, 0, null, localExecutor);
+      RetryingFuture<String> future = executor.createFuture(callable);
+
+      assertNull(future.peekAttemptResult());
+      assertSame(future.getAttemptResult(), future.getAttemptResult());
+      assertFalse(future.getAttemptResult().isDone());
+      assertFalse(future.getAttemptResult().isCancelled());
+
+      executor.submit(future);
+
+      CustomException exception;
+      CancellationException cancellationException = null;
+      int checks = 0;
+      int failedCancelations = 0;
+      do {
+        exception = null;
+        checks++;
+        Future<String> attemptResult = future.getAttemptResult();
+        try {
+          attemptResult.get();
+          assertNotNull(future.peekAttemptResult());
+        } catch (CancellationException e) {
+          cancellationException = e;
+        } catch (ExecutionException e) {
+          exception = (CustomException) e.getCause();
+        }
+        assertTrue(attemptResult.isDone());
+        if (!future.cancel(true)) {
+          failedCancelations++;
+        }
+      } while (exception != null && checks < maxRetries);
+
+      assertTrue(future.isDone());
+      assertNotNull(cancellationException);
+      // future.cancel(true) may return false sometimes, which is ok. Also, the every cancellation of
+      // an already cancelled future should return false (this is what -1 means here)
+      assertEquals(2, checks - (failedCancelations - 1));
+      assertTrue(future.getAttemptSettings().getAttemptCount() > 0);
+      assertFutureCancel(future);
+      localExecutor.shutdownNow();
+    }
+  }
+
+  @Test
+  public void testCancelOuterFutureAfterStart() throws Exception {
+    for (int executionsCount = 0; executionsCount < EXECUTIONS_COUNT; executionsCount++) {
+      ScheduledExecutorService localExecutor = Executors.newSingleThreadScheduledExecutor();
+      FailingCallable callable = new FailingCallable(4, "SUCCESS");
+      RetrySettings retrySettings =
+          FAST_RETRY_SETTINGS
+              .toBuilder()
+              .setInitialRetryDelay(Duration.ofMillis(1_000L))
+              .setMaxRetryDelay(Duration.ofMillis(1_000L))
+              .setTotalTimeout(Duration.ofMillis(10_0000L))
+              .build();
+      RetryingExecutor<String> executor =
+          getRetryingExecutor(retrySettings, 0, null, localExecutor);
+      RetryingFuture<String> future = executor.createFuture(callable);
+      executor.submit(future);
+
+      Thread.sleep(30L);
+
+      boolean res = future.cancel(false);
+      assertTrue(res);
+      assertFutureCancel(future);
+      assertTrue(future.getAttemptSettings().getAttemptCount() < 4);
+      localExecutor.shutdownNow();
+    }
+  }
+
+  @Test
+  public void testCancelProxiedFutureAfterStart() throws Exception {
+    // this is a heavy test, which takes a lot of time, so only few executions.
```

Comment:
Just around one second per iteration on my machine, so currently 2 seconds for the test (since there are only 2 iterations). Before commiting I usually check that test with 1000 iterations per test (and yet somehow they manage to fail on travis =(, probably because travis build machine is too much slower than google workstations).